policy-controller/k8s/status/src/index.rs

```diff
@@ -366,28 +363,39 @@ impl Index {
         // Clone the claims watch out of the index. This will immediately
         // drop the read lock on the index so that it is not held for the
         // lifetime of this function.
-        let mut claims = index.read().claims.clone();
+        let (local, mut claims) = {
+            let idx = index.read();
+            (idx.name.clone(), idx.claims.clone())
+        };
+
+        let mut timer = time::interval(reconciliation_period);
+        timer.set_missed_tick_behavior(time::MissedTickBehavior::Delay);
 
+        let mut is_leader = claims.borrow_and_update().is_current_for(&local);
+
+        // Update our leadership status as the claim changes, triggering an
+        // immediate reconciliation when we become leader. When we are not the
+        // leader, we do not consult the reconciliation timer.
         loop {
             tokio::select! {
+                biased;
+
                 res = claims.changed() => {
                     res.expect("Claims watch must not be dropped");
-                    tracing::debug!("Lease holder has changed");
+                    let claim = claims.borrow_and_update();
+                    tracing::trace!(?claim, "Changed");
+                    let was_leader = is_leader;
+                    is_leader = claim.is_current_for(&local);
+                    if is_leader && !was_leader {
+                        tracing::debug!("Became leader; resetting timer");
+                        timer.reset_immediately();
+                    }
                 }
-                _ = time::sleep(reconciliation_period) => {}
-            }
 
-            // The claimant has changed, or we should attempt to reconcile all
-            //routes to account for any errors. In either case, we should
-            // only proceed if we are the current leader.
-            let claims = claims.borrow_and_update();
-            let index = index.read();
-
-            if !claims.is_current_for(&index.name) {
-                continue;
+                _ = timer.tick(), if is_leader => {
+                    index.read().reconcile();
+                }
```

Comment:
It's perhaps worth separating this reworking into its own PR, but the net result is, I hope, an easier to reason about reconciliation loop.